### PR TITLE
Authentication & Spring Security Impl

### PR DESCRIPTION
**Pull request by akobirswe**

### CODE REVIEW

File: service/src/main/java/com/epam/esm/service/impl/RoleServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+@Service
+@Transactional
+public class RoleServiceImpl implements RoleService {
+    private final RoleRepositoryImpl roleRepository;
+
+    public RoleServiceImpl(RoleRepositoryImpl roleRepository) {
+        this.roleRepository = roleRepository;
+    }
+
+    @Override
```
Comment (sherzodbeek) [inline service/src/main/java/com/epam/esm/service/impl/RoleServiceImpl.java]: You can use inheritDoc to get javadoc from interface to its implementation

File: web/src/main/java/com/epam/esm/controller/AuthController.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+@RequestMapping(value = "/api/auth", produces = "application/json", consumes = "application/json")
+public class AuthController {
+    private final UserService userService;
+
+    public AuthController(UserService userService) {
+        this.userService = userService;
+    }
+
+    @PostMapping("/generate-token")
+    public ResponseEntity<String> generateToken(@Valid @RequestBody TokenRequest request) {
```
Comment (sherzodbeek) [inline web/src/main/java/com/epam/esm/controller/AuthController.java]: Maybe it would be better if this endpoint is called as **login** not **generate-token**

File: web/src/main/java/com/epam/esm/controller/GiftCertificateController.java [hunk 3]
```diff
@@ -27,12 +28,14 @@ public GiftCertificateController(GiftCertificateService giftCertificateService,
     }
 
     @GetMapping(value = "/{id}")
+    @PreAuthorize("hasAnyAuthority('ROLE_USER','ROLE_ADMIN')")
```
Comment (sherzodbeek) [inline web/src/main/java/com/epam/esm/controller/GiftCertificateController.java]: Seems to me you missed this requirement: 
- Guest:
    * Read operations for main entity.
    
It means that Guest users should be able to see the gift certificates

**Comment by sherzodbeek** [comment]
@AkobirToshtemirov I have left my comments. I have not seen tests for new services. Please add them

File: web/src/main/java/com/epam/esm/controller/GiftCertificateController.java (hunk 3 above)
Comment (akobirswe) [reply web/src/main/java/com/epam/esm/controller/GiftCertificateController.java]: I got you. I will fix it.

**Comment by AkobirToshtemirov** [comment]
What about JWT and Security part. Do these parts cover requirements?

Thanks.

File: service/src/test/java/com/epam/esm/UserServiceTest.java [hunk 4]
```diff
@@ -48,4 +68,24 @@ void testFindUserById() {
...
+    void testGenerateToken() {
+        TokenRequest tokenRequest = new TokenRequest("testUser", "password");
+        when(authenticationManager.authenticate(any(UsernamePasswordAuthenticationToken.class))).thenReturn(null);
+        when(jwtTokenUtil.generateToken(anyString())).thenReturn(new TokenResponse("accessToken"));
+        TokenResponse tokenResponse = userService.generateToken(tokenRequest);
+
+        assertEquals("accessToken", tokenResponse.accessToken());
+        verify(authenticationManager, times(1)).authenticate(any(UsernamePasswordAuthenticationToken.class));
+        verify(jwtTokenUtil, times(1)).generateToken(anyString());
+    }
```
Comment (sherzodbeek) [inline service/src/test/java/com/epam/esm/UserServiceTest.java:91]: Could you please add tests for negative cases also? For example, username may be already exist or the password can be wrong.

File: web/src/main/java/com/epam/esm/controller/RoleController.java [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+public class RoleController {
+    private final RoleServiceImpl roleService;
+
+    public RoleController(RoleServiceImpl roleService) {
+        this.roleService = roleService;
+    }
+
+    @PostMapping()
+    @PreAuthorize("hasRole('ADMIN')")
+    public ResponseEntity<Role> saveTag(@RequestBody Role role) {
```
Comment (sherzodbeek) [inline web/src/main/java/com/epam/esm/controller/RoleController.java]: Seems to me, method name is wrong here. It should be saveRole.
My question is why do you need API to create new roles? I think it is useless. Imagine you create new role, but how are you going to give permission for this role?

**Review by sherzodbeek — APPROVED** [review]
LGTM
